### ISSUE TITLE
Feature Request: Autodiscovery for Text, Number and Select Devices

### ISSUE THREAD

**Issue by akamming** [domoticz/domoticz#6217]
Hi, 

I noticed several sensors that are published in the MQTT Discovery topics which were not recognized by domoticz (and did a check, they are recognized by HASS as valid sensors)

I analyzed and noticed the following not supported in Domoticz:
- [MQTT Select](https://www.home-assistant.io/integrations/select.mqtt/)
- [MQTT Number](https://www.home-assistant.io/integrations/number.mqtt/)
- and [MQTT Text](https://www.home-assistant.io/integrations/text.mqtt/)

It would be really great if these were  supported by domoticz as well...

**Comment by waltervl** [comment]
You are not referring to MQTT AD devices.

Number is auto detected and visible in the Setup page on MQTT AD  hardware https://wiki.domoticz.com/MQTT#MQTT_Autodiscovery_Setup

Text should also be detected when using the correct config statement using sensor with unit 'text'

If you want to use a random MQTT protocol use the MQTTmapper python plugin.

**Comment by waltervl** [comment]
Select should also be discovered: https://github.com/domoticz/domoticz/blob/bc620cdb67be1562a0c5239e8c8df8a9b86d94bb/hardware/MQTTAutoDiscover.cpp#L1170

Better use the forum to discuss the config statements you tried as it probably fails there.

**Comment by akamming** [comment]
it is really frustrating you just close this "issue" as i try to give feedback to enhance domoticz and then you just close the issue for the wrong reasons.

This is not about failed config messages. The config messages are fine. These devices work perfectly in HASS, it's just that domoticz does not detect them, cause apparently it misinterprets some messages

For example this text sensor is not detected by domoticz, rendering the VolvoAAOS2MQTT useless for domoticz: 

`{"name": "Volvo OTP", "object_id": "volvo_otp", "schema": "state", "command_topic": "volvoAAOS2mqtt/otp_code", "state_topic": "volvoAAOS2mqtt/otp_code/state", "unique_id": "volvoAAOS2mqtt_otp", "pattern": "\\d{6}", "icon": "mdi:two-factor-authentication", "mode": "text"}`

BTW I already tried adding "unit": "text"  to the discovery message, but then it is still not detected.

(and it is a good tip that hardware page, i confirmed the numbers indeed work)

**Comment by waltervl** [comment]
As indicated in the start of a github issue feature request officially should be discussed on the forum first. Second, the requested MQTT AD device types should already be detected. So I assumed you did not investigate properly. Also the given information was too limited.

For issues with MQTT AD we  for a long time have an instruction how to report issues: https://wiki.domoticz.com/MQTT_AD_Report_Problems
Your issue also did not follow that rule. 

Also that HA is discovering this device will not assure that Domoticz will as Domoticz for example is not discovering complex value topics. 
And I suppose for the text device it should be a device type sensor that is supported. I don't see a text type device in MQTT AD (but I am not a developer.....) 

Better start a new Github issue about the text device after you have investigated some more options and give full feedback.

**Comment by akamming** [comment]
Nope, i spent enough time on this.

I already did the steps of ad discovery troubleshooting and provided the requested discovery message. 

I also already referenced the mqtt text device. 

I Will continue using ha for this. If you want to implement in domoticz you have my Full support

but closing issues like this without a developers looking at my findings makes the process too random for me…

**Comment by waltervl** [comment]
As a developer did not have to look at your request as it did not seem to be a bug or a real feature request.

**Comment by gizmocuz** [comment]
@akamming , i know it's dark outside... but let's try to get into Christmas spirit please ;)

Your 'text' node is not recognized because

"mode": "text"

is not recognized by Domoticz, is this a real config or one made by yourself?

Try

"unit": "text"

and it should work

As mentioned by Walter, number types are placed in the hardware page (setup button for your MQTT-AD hardware), as these are for configuration

And 'select' should work.. So, post the complete config topic name and payload so this can be checked

There is a wiki page on how to report MQTT-AD issues, and it involves complete config topic names/path and payload at the least.
Else there is no way for anyone to debug whats causing the issue

**Comment by akamming** [comment]
Tx for the reaction.

This is a real config (from the Volvo2mqtt docker container by dielee). 

If i change the mode to unit. It still is not recognized:
`{"name": "Volvo OTP", "object_id": "volvo_otp", "schema": "state", "command_topic": "volvoAAOS2mqtt/otp_code", "state_topic": "volvoAAOS2mqtt/otp_code/state", "unique_id": "volvoAAOS2mqtt_otp", "pattern": "\\d{6}", "icon": "mdi:two-factor-authentication", "unit": "text"}
`
I would also not expect this to work, cause unit is not a valid tag for a mqtt text  sensor. mode however is (in that tag you can specify wheter it should be normal text or a password). This is all very well documented in the [official mqtt text sensor documentation](https://www.home-assistant.io/integrations/text.mqtt/)

To give you the full set of data you need:

### For the text sensor: 
- Discovery topic: `homeassistant/text/volvoAAOS2mqtt/volvo_otp/config`
- Discovery payload: `{"name": "Volvo OTP", "object_id": "volvo_otp", "schema": "state", "command_topic": "volvoAAOS2mqtt/otp_code", "state_topic": "volvoAAOS2mqtt/otp_code/state", "unique_id": "volvoAAOS2mqtt_otp", "pattern": "\\d{6}", "icon": "mdi:two-factor-authentication", "mode": "text"}`
- State topic: `volvoAAOS2mqtt/otp_code/state`
- state payload: `000000` 

### For the Selector:
(this example is my own creation. But is according to the [official mqqt select sensor documentation](https://www.home-assistant.io/integrations/select.mqtt/) and works quite well in HASS)  
- Discovery Topic: `homeassistant/select/domesphelper/Curvature/config`
- Discovery payload: `{"name":"Curvature","unique_id":"domesphelper_Curvature","stat_t":"domesphelper/select/Curvature/state","cmd_t":"domesphelper/select/Curvature/set","platform":"select","options":["none","small","medium","large","extralarge"],"dev":{"ids":"E8DB8498C1DF","name":"domesphelper","sw":"domesphelper_Nov 27 2024_17:14:01","mdl":"d1_mini","mf":"espressif","cu":"http://192.168.2.6/"}}`
- State Topic: `domesphelper/select/Curvature/state`
- State Payload: `small`

Hope this helps..

**Comment by waltervl** [comment]
For the text device instead if config topic `homeassistant/text/volvoAAOS2mqtt/volvo_otp/config` 
try 
`homeassistant/sensor/volvoAAOS2mqtt/volvo_otp/config`

**Comment by akamming** [comment]
> For the text device instead if config topic `homeassistant/text/volvoAAOS2mqtt/volvo_otp/config` try `homeassistant/sensor/volvoAAOS2mqtt/volvo_otp/config`

This would not make sense: altough this would be a valid discovery message. This is the topic for custom sensors instead of a text sensors.  

To rule out any dicussion, i tried anyway using mqtt explorer, and confirmed that this creates a custom sensor instead of a text device: 
![image](https://github.com/user-attachments/assets/3d37b7b6-4a11-449d-a4e9-f8ec7633df39)

**Comment by waltervl** [comment]
Hmm , you also used "unit": "text" in the config payload? Then gizmocuz informed me incorrectly....

**Comment by akamming** [comment]
hmm.. a bit of wrong copying and pasting the wrong discovery message, so corrected in the previous commet ...
 i very likely used the tag "mode": "text", but have to verify when i have access to my pi again... 

I don't expect it to do anything, cause "unit" is not an officialy described tag in that descory message, but let me verify that...

**Comment by gizmocuz** [comment]
unit is described, and it is valid to create a 'sensor' object, where the unit is 'text'.

But I also got the Text device working with the sample above. Do you also need it to be edited? This is something I still have to implement because currently text objects are status objects

About your own selector switch, try the below example (and adjust to your liking)

```
{
  "stat_t": "~/ess_mode/state",
  "dev": {
    "manufacturer": "PA1DVB",
    "ids": [
      "0x8000000001"
    ]
  },
  "~": "my_victron",
  "name": "ESS Control Mode",
  "options": [
    "Mode 1",
    "Mode 2"
  ],
  "cmd_t": "~/set/ess_mode",
  "uniq_id": "0x8000000001",
  "val_tpl": "{{ value_json.ess_mode }}"
}
```

**Comment by gizmocuz** [comment]
Could you try beta 16346

This version should give you working "text" type sensors, and you should be able to edit the text sensors and when pressing OK, the data is send to the command topic in MQTT

**Comment by akamming** [comment]
> Could you try beta 16346
> 
> This version should give you working "text" type sensors, and you should be able to edit the text sensors and when pressing OK, the data is send to the command topic in MQTT

Nice work!!!!!

Did a quick test: 
- The text type sensor is indeed created and is updated if the device sends updated values. 
- The "aanpassen" button is also available, and i can enter different texts
- However the behaviour is incorrect. It sends to the State Topic instead of the Command Topic...

**Comment by akamming** [comment]
tomorrow i have more time, and will then check the select discovery message...

**Comment by akamming** [comment]
Just analyzed the selector disovery message you (@gizmocuz ) provided. The main difference with "mine" is that a value_template was provided. I just added that to my discovery message and now it's working fine. Tx...!

So that leaves only the updating of a text sensor (which now publishes a payload on the state_topic instead of the command_topic as described above).

**Comment by gizmocuz** [comment]
`However the behaviour is incorrect. It sends to the State Topic instead of the Command Topic...`

@akamming , whoops, sorry about this. This is now corrected in beta 16348

About the selector switch, I think a value template is only necessary when your data resides in a JSON payload.

I just checked your config above, and it creates a selector switch without issues:

![image](https://github.com/user-attachments/assets/14ad52a6-d284-4557-81a8-664a77f598ad)

When sending 'small' to the state topic, the selector switch changes its state to 'small', so no issues here as well

Switching in Domoticz to 'large' also sends 'large' to the set MQTT topic, so no issues here as well

There was no problem with the 'Select' switch

Hope the text object now works, if yes, please close this issue

**Comment by akamming** [comment]
Wow you're fast...

I just checked, and indeed String selector now updates the correct topic...

I also just rechecked the selector switch. It is vague, cause indeed now it is also created without value template. 

The only difference is that the first time i tested the selector with value template i already updated to latest beta. So maybe there's a difference.

Anyway it works now.  Thanx for the help and the good addition of MQTT Text sensors.

Keep up the good work, closing the issue!
